### ISSUE TITLE
PTT: Point-Trajectory Transformer for Efficient Temporal 3D Object Detection

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#62]
### Paper Link
https://openaccess.thecvf.com/content/CVPR2024/html/Huang_PTT_Point-Trajectory_Transformer_for_Efficient_Temporal_3D_Object_Detection_CVPR_2024_paper.html

### Code Link
https://github.com/kuanchihhuang/PTT

## Task Definiation: Multi-frame/Temporal LiDAR-based 3D Object Detection

- leverages point cloud **sequences** captured over time as the vehicle moves in real-world scenarios.
- offer a more comprehensive and denser representation of the surrounding environment.
- one straightforward method to tackle sequential data is **concatenating all multi-frame LiDAR as input**, followed by the original 3D detection pipeline, which provides performance benefit via the denser information.
- recent methods typically adopt a two-stage approach: 1. They generate bounding **box candidates** and select **objects’ point clouds** using region proposal networks (RPN). 2. These approaches **concatenate point features with proposal features** for 3D object detection. 

<img width="500" alt="image" src="https://github.com/user-attachments/assets/c6dfe797-c6ca-4b37-be4f-bc3b5bb1352b">

### Problems 
1. Sampling object point clouds for each frame in sequential modeling causes a **memory burden** for a long sequence during the training and testing phases.
2. These methods neglects the significance of the **interaction** between point and proposal features.

## Methodology

<img width="1626" alt="image" src="https://github.com/user-attachments/assets/25d1e1a8-1290-4a49-b5e9-e964dafe9cba">

### Point-Trajectory Feature Generation

**Point-to-Proposal Features**: We first generate **point-to-proposal features** based on the offset between points and points of the bounding box (i.e., center and eight corners):

$$
G^t = \text{MLP}(\text{Concat}([I - b_i^t]_{i=0}^8, \Delta t)) \in \mathbb{R}^{N \times C},
$$

where $N$ is the number of the sampled points, $C$ is the feature dimension, and $\Delta t$ is the time difference between $b^t$ to the current frame. Then $\{ G^t \}_{1}^{T}$ denotes the sequential Point-to-Proposal features that encode the point features based on its proposal at different frame $t$. To this end, we are **not required to store each frame’s point cloud** for the object but can **still preserve its geometric and motion features related to previous frames**.

**Proposal-to-Proposal Features.** By considering proposals from the previous $T$ frames, we can encode the relative location of the proposals compared to the current frame, enhancing the model’s ability to capture the motion pattern more effectively. The Proposal-to-Proposal features can be defined as:

$$
F^t = \text{MLP}(\text{Concat}(B_{xyz}^t - B_{xyz}^T, B_{whl\theta}^t, \Delta t)) \in \mathbb{R}^{C}
$$

where $B_{xyz}$ and $B_{whl\theta}$ represent the attributes of the proposal, including the center location $(x, y, z)$ and the size of the bounding box $(w, l, h)$ respectively, along with the heading $\theta$. To this end, $\{F^t\}_{1}^{T}$ is the sequential Proposal-to-Proposal features.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/44eedf6b-a93c-440d-b650-fa3d8530571d">

**Sequential Point-Trajectory Features.** After deriving the Point-to-Proposal and Proposal-to-Proposal features from [(1)](#) and [(2)](#), we can combine them into integrated features. Specifically, we **compress the per-frame Point-to-Proposal features using a MaxPool operation** and **concatenate them with the corresponding Proposal-to-Proposal features** to acquire per-frame point-trajectory features:

$$
P^t = \text{MLP}(\text{Concat}(F^t, \text{MaxPool}(G^t))).
$$

In this manner, we can use $\{P^t\}_{1}^{T}$ to represent the sequential point-trajectory information, which is utilized for further feature learning.

### Point-Trajectory Transformer (PTT)

After acquiring the sequential point-trajectory features, our objective is to devise a point-trajectory transformer with long short-term memory to encode these features. Our transformer comprises four modules: 1) long-term memory, 2) short-term memory, 3) future-aware encoder, and 3) point-trajectory aggregator module.

**Long-Term Memory.** 
- Contrary to prior approaches that may be limited to handling short sequences, our memory-efficient solution can manage **significantly longer sequences**. 
- We partition the feature sequence into long-term and short-term memories to capture sequential information effectively across various periods. 
- Considering the total frame length $T$, we separate the sequential point-trajectory features $\{P^t\}$ into the short memory as $M_s = \{P^T, ..., P^{T-m_s+1}\}$ with a short frame number $m_s$, and $M_l = \{P^{m_l}, ..., P^1\}$ with a long frame number $m_l$, where $m_l + m_s = T$.

For the long-term memory $M_l \in \mathbb{R}^{m_l \times C}$, we utilize the multi-head self-attention encoder layers to encode the long-term features:

$$
\mathbf{Q} = M_l W_q, \quad \mathbf{K} = M_l W_k, \quad \mathbf{V} = M_l W_v,
$$

$$
\hat{M}_l = \text{FFN}(\text{MultiHead}(\mathbf{Q}, \mathbf{K}, \mathbf{V})),
$$

where $W_*$ denotes the learnable parameters for the long-term memory. We utilize one linear layer followed by the ReLU activation to construct our feed-forward network FFN. To this end, we can obtain the aggregated long-term trajectory features $\hat{M}_l$.

**Short-Term Memory.** 
- Short-term features are crucial for learning 3D bounding boxes in the current frame, as they capture the dynamic status close to the current frame. - We use the short-term memory $M_s$ as a query to extract valuable trends in historical sequences from the aggregated long-term memory $\hat{M}_l$. 
- We exploit the transformer module to model the **relationship between short- and long-term memories**:

$$
\mathbf{Q}^s = M_s W_q^s, \quad \mathbf{K}^l = \hat{M}_l W_k^l, \quad \mathbf{V}^l = \hat{M}_l W_v^l,
$$

$$
\hat{M}_s = \text{FFN}(\text{MultiHead}(\mathbf{Q}^s, \mathbf{K}^l, \mathbf{V}^l)),
$$

where $\hat{M}_s$ denotes the **interactive short-term memory features**. 

<img width="500" alt="image" src="https://github.com/user-attachments/assets/03882231-0cec-4195-b535-796a3cfa2cbb">

**Future-Aware Activation.**

- Future information is also important for 3D estimation in the current frame by improving feature learning.
- Considering a 3D proposal box with $(x, y, z, w, l, h, \theta)$ with the estimated velocity $(v_x, v_y)$ at frame $T$, we can infer the box’s state at future time difference $\Delta t$ as $\mathbf{B}^{T + \Delta t}$:

$$
(x + \Delta t \cdot v_x, y + \Delta t \cdot v_y, z, w, l, h, \theta + \Delta t \cdot v_{\theta}),
$$

where $v_{\theta}$ can be calculated by the difference within previous frames: $v_{\theta} = \frac{\theta_T - \theta_1}{T}.$

- We derive the future-aware features $M_f = \{P^{T+m_f}, ..., P^{T+1}\}$ for $m_f$ future frames based on the inferred future box $\{\mathbf{B}^{T+\Delta t}_t\}_{t=1}^{m_f}$. 
- We then concatenate the long-term $\hat{M}_l$ and the short-term memory $\hat{M}_s$ to generate the whole memory features $M = [\hat{M}_l, \hat{M}_s]$, where $[\cdot, \cdot]$ means the feature concatenation operation along the temporal dimension. 
- Afterward, since the inferred future boxes may be **inaccurate**, we concatenate the learnable queries $\mathbf{q}_f \in \mathbb{R}^{m_f \times C}$ to the future features along the channel dimension, followed by the convolution operation:

$$
M_f' = \text{Conv}(\text{Concat}(\mathbf{q}_f, M_f)),
$$

facilitating additional learning for potential correction. Then the concatenated features $M_f'$ are fed to the multi-head self-attention encoder layers:

$$
\mathbf{Q}^f = M_f' W_q^f, \quad \mathbf{K}^f = M W_k^f, \quad \mathbf{V}^f = M W_v^f,
$$

$$
\hat{M}_f' = \text{FFN}(\text{MultiHead}(\mathbf{Q}^f, \mathbf{K}^f, \mathbf{V}^f)),
$$

where $\hat{M}_f'$ is the future-aware features.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/be5ceb24-5e77-4788-a250-43326f47e68a">

**Point-Trajectory Aggregator.**

After obtaining the enhanced sequential point-trajectory features based on long-term, short-term, and future encoding, we integrate them with the current frame’s point features into the proposed point-trajectory aggregator as shown in Figure 3.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/2ef034c8-e24b-4413-8936-15b48eec2b4b">

## Implementation 

- We utilize the CenterPoint [33] as our region proposal network (RPN) to generate first-stage high-quality 3D proposals. 
- Building upon these proposals, we train our proposed PTT for 6 epochs, utilizing the ADAM optimizer with an initial learning rate of 0.003 and a batch size of 4. 
- We set an IoU matching threshold of 0.5 for generating the proposed trajectories in the first stage. 
- Also, we apply proposal-centric box jitter augmentation to the per-frame 3D proposal box, following the approach outlined in PointRCNN.
- Each proposal is randomly sampled with 128 raw LiDAR points. 
- The feature dimension for point clouds and trajectories is configured as 128. 
- All the experiments are implemented in Pytorch on an NVIDIA 3090 GPU. 
- We conduct ablation studies by using a light-cost setting of training our PTT with 32 frames for 3 epochs.

## Experiments 
<img width="1473" alt="image" src="https://github.com/user-attachments/assets/e9de6fc2-fced-4f13-9bbc-07539ca62b84">

**Runtime Performance.** 

- **MPPNet** takes approximately 1100ms to process 16-frame sequential data, our PTT only requires about 150ms to run 64-frame information on the same devices.
- When comparing our approach using 64 frames with **MSF** using 8 frames, our method also achieves better performance across most metrics. It is worth mentioning that, although our PTT exhibits a comparable runtime to MSF (approximately 160ms), their approach necessitates retaining the entire sequential point cloud to propagate the current frame’s proposals into previous frames. This requirement results in a larger memory bank, potentially causing memory overhead and limiting its applicability to longer sequences.

### More Ablation

<img width="500" alt="image" src="https://github.com/user-attachments/assets/c93f7563-8e57-4d28-abf4-aa1d05c7d25a">

<img width="500" alt="image" src="https://github.com/user-attachments/assets/6dbfcbb2-c65a-4361-a153-3fb22b38561b">
